corehq/apps/userreports/pillow.py

```diff
@@ -249,6 +254,32 @@ def rebuild_table(self, adapter, diffs=None):
         else:
             adapter.rebuild_table(source='pillowtop')
 
+    def _pull_in_new_and_modified_data_sources(self):
+        """
+        Find any data sources that have been modified since the last time this was bootstrapped
+        and update the in-memory references.
+        """
+        new_last_imported = datetime.utcnow()
```

Comment:
Is there a race condition where the last modified date is set on a data source just before this date is calculated, but that data source is not fully saved until after the `get_data_sources_modified_since` returns? If that happens then that data source will slip through the crack since its last modified date is older than `new_last_imported`?

Is this the race condition you were speculating about on the call?

Reply:
(not sure off the top of my head and not the race condition i had in mind)

Reply:
Sorry about the rushed response yesterday. I do think what you're describing is plausible although also suspect there are many such types of race conditions that might exist anywhere we have logic that tries to get anything modified since a checkpointed date (though perhaps that's a rare thing to do in HQ).

In this scenario the data source would get picked up on the next `bootstrap` call (currently 3 hours later) so basically fall back to the previous behavior. That said, if we do decide to remove the `bootstrap` stuff entirely, then it would be much worse.

Do you have any idea how you'd mitigate such a thing? I think whenever dates are being set in app-level code then shipped to the DB you'd have this type of problem.

Reply:
> Do you have any idea how you'd mitigate such a thing? I think whenever dates are being set in app-level code then shipped to the DB you'd have this type of problem.

Yeah, using dates as sequence checkpoints is not a great practice. Could use a [monotonically increasing sequence](https://en.wikipedia.org/wiki/Sequence#Increasing_and_decreasing) ([postgres has them](https://www.postgresqltutorial.com/postgresql-sequences/)).

Reply:
That makes sense. Thanks.

Reply:
Thinking about this more, I may be wrong. A monotonically increasing sequence might not be much better (although it does provide discrete values). I think I'm coming to the conclusion that real change feeds are a non-trivial to implement.